monai/apps/generation/maisi/networks/controlnet_maisi.py

```diff
@@ -0,0 +1,158 @@
+    Diffusion Models" (https://arxiv.org/abs/2302.05543)
+
+    Args:
+        spatial_dims: number of spatial dimensions.
+        in_channels: number of input channels.
+        num_res_blocks: number of residual blocks (see ResnetBlock) per level.
+        num_channels: tuple of block output channels.
+        attention_levels: list of levels to add attention.
+        norm_num_groups: number of groups for the normalization.
+        norm_eps: epsilon for the normalization.
+        resblock_updown: if True use residual blocks for up/downsampling.
+        num_head_channels: number of channels in each attention head.
+        with_conditioning: if True add spatial transformers to perform conditioning.
+        transformer_num_layers: number of layers of Transformer blocks to use.
+        cross_attention_dim: number of context dimensions to use.
+        num_class_embeds: if specified (as an int), then this model will be class-conditional with `num_class_embeds`
+            classes.
+        upcast_attention: if True, upcast attention operations to full precision.
+        use_flash_attention: if True, use flash attention for a memory efficient attention mechanism.
+        conditioning_embedding_in_channels: number of input channels for the conditioning embedding.
+        conditioning_embedding_num_channels: number of channels for the blocks in the conditioning embedding.
+    """
+
+    def __init__(
+        self,
+        spatial_dims: int,
+        in_channels: int,
+        num_res_blocks: Sequence[int] | int = (2, 2, 2, 2),
+        num_channels: Sequence[int] = (32, 64, 64, 64),
+        attention_levels: Sequence[bool] = (False, False, True, True),
+        norm_num_groups: int = 32,
+        norm_eps: float = 1e-6,
+        resblock_updown: bool = False,
+        num_head_channels: int | Sequence[int] = 8,
+        with_conditioning: bool = False,
+        transformer_num_layers: int = 1,
+        cross_attention_dim: int | None = None,
+        num_class_embeds: int | None = None,
+        upcast_attention: bool = False,
+        use_flash_attention: bool = False,
+        conditioning_embedding_in_channels: int = 1,
+        conditioning_embedding_num_channels: Sequence[int] | None = (16, 32, 96, 256),
+    ) -> None:
+        super().__init__(
+            spatial_dims,
+            in_channels,
+            num_res_blocks,
+            num_channels,
+            attention_levels,
+            norm_num_groups,
+            norm_eps,
+            resblock_updown,
+            num_head_channels,
+            with_conditioning,
+            transformer_num_layers,
+            cross_attention_dim,
+            num_class_embeds,
+            upcast_attention,
+            use_flash_attention,
+            conditioning_embedding_in_channels,
+            conditioning_embedding_num_channels,
+        )
+
+
+    def forward(
```

Comment:
I would suggest splitting the forward method into smaller functional chunks to make it easier to understand and maintain. What do you think?
Something like this: 
```
class CustomControlNet(ControlNet):
    def __init__(self, ...):
        super().__init__(...)

    def forward(self, x, timesteps, controlnet_cond, conditioning_scale=1.0, context=None, class_labels=None):
        emb = self._prepare_time_and_class_embedding(timesteps, class_labels)
        h = self._initial_convolution(x)
        h += self.controlnet_cond_embedding(controlnet_cond)
        h = self._apply_down_blocks(emb, context, h)
        h = self._apply_mid_block(emb, context, h)
        result = self._apply_controlnet_blocks(h, conditioning_scale)
        return result

    def _prepare_time_and_class_embedding(self, timesteps, class_labels):
        ...

    def _initial_convolution(self, x):
        ...

    def _apply_down_blocks(self, emb, context, h):
        ...
        return down_block_res_samples, mid_block_res_sample

    def _apply_mid_block(self, emb, context, h):
        ...

    def _apply_controlnet_blocks(self, h, conditioning_scale):
        ...
        return down_block_res_samples, mid_block_res_sample
```

Reply:
Thanks, refactored forward func.